From: Chad Monroe <chad@monroe.io>
Date: Fri, 15 May 2026 13:45:09 -0700
Subject: [PATCH] net: ethernet: mtk_ppe: support MXL862 8021Q DSA tag

mtk_flow_get_dsa_port() hard-coded DSA_TAG_PROTO_MTK, dropping every
MXL86282 user-port flow to the software path. Admit
DSA_TAG_PROTO_MXL862_8021Q and pass the protocol to
mtk_foe_entry_set_dsa(), which encodes the source port in VLAN bits
[11:10] via mtk_foe_entry_set_vlan() for MXL862 versus l2.etype for the
native MTK tag.

Based-on-patch-by: chak-kei.lam <chak-kei.lam@mediatek.com>
Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/ethernet/mediatek/mtk_ppe.c         |   32 +++++++++++++++++-------
 drivers/net/ethernet/mediatek/mtk_ppe.h         |    2 -
 drivers/net/ethernet/mediatek/mtk_ppe_offload.c |   15 +++++++----
 3 files changed, 34 insertions(+), 15 deletions(-)

--- a/drivers/net/ethernet/mediatek/mtk_ppe.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.c
@@ -382,20 +382,34 @@ int mtk_foe_entry_set_ipv6_tuple(struct
 }
 
 int mtk_foe_entry_set_dsa(struct mtk_eth *eth, struct mtk_foe_entry *entry,
-			  int port)
+			  int proto, int port)
 {
-	struct mtk_foe_mac_info *l2 = mtk_foe_entry_l2(eth, entry);
+#if IS_ENABLED(CONFIG_NET_DSA)
+	struct mtk_foe_mac_info *l2;
 
-	l2->etype = BIT(port);
+	if (proto == DSA_TAG_PROTO_MXL862_8021Q) {
+		/* MXL862 8021Q tag carries the source port in VLAN bits
+		 * [11:10] of an otherwise zero VID. mtk_foe_entry_set_vlan()
+		 * already programs the FOE vlan layer/tag fields correctly
+		 * for this encoding.
+		 */
+		mtk_foe_entry_set_vlan(eth, entry, port + GENMASK(11, 10));
+	} else {
+		l2 = mtk_foe_entry_l2(eth, entry);
+		l2->etype = BIT(port);
+
+		if (!(entry->ib1 & mtk_get_ib1_vlan_layer_mask(eth)))
+			entry->ib1 |= mtk_prep_ib1_vlan_layer(eth, 1);
+		else
+			l2->etype |= BIT(8);
 
-	if (!(entry->ib1 & mtk_get_ib1_vlan_layer_mask(eth)))
-		entry->ib1 |= mtk_prep_ib1_vlan_layer(eth, 1);
-	else
-		l2->etype |= BIT(8);
-
-	entry->ib1 &= ~mtk_get_ib1_vlan_tag_mask(eth);
+		entry->ib1 &= ~mtk_get_ib1_vlan_tag_mask(eth);
+	}
 
 	return 0;
+#else
+	return -EOPNOTSUPP;
+#endif
 }
 
 int mtk_foe_entry_set_vlan(struct mtk_eth *eth, struct mtk_foe_entry *entry,
--- a/drivers/net/ethernet/mediatek/mtk_ppe.h
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.h
@@ -393,7 +393,7 @@ int mtk_foe_entry_set_ipv6_tuple(struct
 				 __be32 *src_addr, __be16 src_port,
 				 __be32 *dest_addr, __be16 dest_port);
 int mtk_foe_entry_set_dsa(struct mtk_eth *eth, struct mtk_foe_entry *entry,
-			  int port);
+			  int proto, int port);
 int mtk_foe_entry_set_vlan(struct mtk_eth *eth, struct mtk_foe_entry *entry,
 			   int vid);
 int mtk_foe_entry_set_pppoe(struct mtk_eth *eth, struct mtk_foe_entry *entry,
--- a/drivers/net/ethernet/mediatek/mtk_ppe_offload.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe_offload.c
@@ -169,7 +169,7 @@ mtk_flow_mangle_ipv4(const struct flow_a
 }
 
 static int
-mtk_flow_get_dsa_port(struct net_device **dev)
+mtk_flow_get_dsa_port(struct net_device **dev, int *proto)
 {
 #if IS_ENABLED(CONFIG_NET_DSA)
 	struct dsa_port *dp;
@@ -178,11 +178,15 @@ mtk_flow_get_dsa_port(struct net_device
 	if (IS_ERR(dp))
 		return -ENODEV;
 
-	if (dp->cpu_dp->tag_ops->proto != DSA_TAG_PROTO_MTK)
+	if (dp->cpu_dp->tag_ops->proto != DSA_TAG_PROTO_MTK &&
+	    dp->cpu_dp->tag_ops->proto != DSA_TAG_PROTO_MXL862_8021Q)
 		return -ENODEV;
 
 	*dev = dsa_port_to_conduit(dp);
 
+	if (proto)
+		*proto = dp->cpu_dp->tag_ops->proto;
+
 	return dp->index;
 #else
 	return -ENODEV;
@@ -284,7 +288,7 @@ mtk_flow_set_output_device(struct mtk_et
 {
 	struct mtk_wdma_info info = {};
 	struct mtk_mac *mac;
-	int pse_port, dsa_port, queue;
+	int pse_port, dsa_port, dsa_proto, queue;
 
 	if (mtk_flow_get_wdma_info(dev, dest_mac, &info) == 0) {
 		mtk_foe_entry_set_wdma(eth, foe, info.wdma_idx, info.queue,
@@ -310,7 +314,7 @@ mtk_flow_set_output_device(struct mtk_et
 		goto out;
 	}
 
-	dsa_port = mtk_flow_get_dsa_port(&dev);
+	dsa_port = mtk_flow_get_dsa_port(&dev, &dsa_proto);
 
 	if (dev == eth->netdev[0]) {
 		mac = eth->mac[0];
@@ -326,7 +330,7 @@ mtk_flow_set_output_device(struct mtk_et
 	}
 
 	if (dsa_port >= 0) {
-		mtk_foe_entry_set_dsa(eth, foe, dsa_port);
+		mtk_foe_entry_set_dsa(eth, foe, dsa_proto, dsa_port);
 		queue = MTK_MAX_DEVS + eth->pppq_ofs[mac->id] +
 			mac->dsa_user_idx[dsa_port];
 	} else {
@@ -405,6 +409,11 @@ mtk_flow_offload_replace(struct mtk_eth
 		flow_rule_match_meta(rule, &match);
 		if (mtk_is_netsys_v2_or_greater(eth)) {
 			idev = __dev_get_by_index(&init_net, match.key->ingress_ifindex);
+			/* Called for the side effect of rewriting @idev to its
+			 * DSA conduit when the ingress is a DSA user port; the
+			 * returned port index is not needed here.
+			 */
+			mtk_flow_get_dsa_port(&idev, NULL);
 			if (mtk_flow_is_valid_idev(eth, idev)) {
 				struct mtk_mac *mac = netdev_priv(idev);
 
